lib/load.php

```diff
@@ -24,31 +24,36 @@
 require dirname( __FILE__ ) . '/register.php';
 require dirname( __FILE__ ) . '/demo.php';
 
-// Register server-side code for individual blocks.
-if ( ! function_exists( 'render_block_core_archives' ) ) {
-	require dirname( __FILE__ ) . '/../packages/block-library/src/archives/index.php';
-}
-if ( ! function_exists( 'render_block_core_block' ) ) {
-	require dirname( __FILE__ ) . '/../packages/block-library/src/block/index.php';
-}
-if ( ! function_exists( 'render_block_core_categories' ) ) {
-	require dirname( __FILE__ ) . '/../packages/block-library/src/categories/index.php';
-}
-// Currently merged in core as `gutenberg_render_block_core_latest_comments`,
-// expected to change soon.
-if ( ! function_exists( 'render_block_core_latest_comments' )
-	&& ! function_exists( 'gutenberg_render_block_core_latest_comments' ) ) {
-	require dirname( __FILE__ ) . '/../packages/block-library/src/latest-comments/index.php';
-}
-if ( ! function_exists( 'render_block_core_latest_posts' ) ) {
-	require dirname( __FILE__ ) . '/../packages/block-library/src/latest-posts/index.php';
-}
-if ( ! function_exists( 'render_block_core_rss' ) ) {
-	require dirname( __FILE__ ) . '/../packages/block-library/src/rss/index.php';
-}
-if ( ! function_exists( 'render_block_core_shortcode' ) ) {
-	require dirname( __FILE__ ) . '/../packages/block-library/src/shortcode/index.php';
-}
-if ( ! function_exists( 'render_block_core_search' ) ) {
-	require dirname( __FILE__ ) . '/../packages/block-library/src/search/index.php';
+/**
+ * Discovers block files from the plugin built artifact, unregistering the
+ * equivalent in core if already defined, then includes the block file.
+ */
+function gutenberg_reregister_core_block_types() {
+	// Blocks directory may not exist if working from a fresh clone.
+	$blocks_dir = dirname( __FILE__ ) . '/../build/block-library/blocks/';
+	if ( ! file_exists( $blocks_dir ) ) {
+		return;
+	}
+
+	$registry = WP_Block_Type_Registry::get_instance();
+
+	$files = scandir( $blocks_dir );
+	foreach ( $files as $file ) {
+		$parts = pathinfo( $file );
+
+		// Ignore all non-PHP files, subdirectories, path traversal.
+		if ( 'php' !== $parts['extension'] ) {
+			continue;
+		}
+
+		// Derive the assumed block name by file path.
+		$block_name = 'core/' . $parts['filename'];
+
+		if ( $registry->is_registered( $block_name ) ) {
```

Comment:
I was wondering why it isn't using `remove_action` against `add_action( 'init', 'register_block_core_latest_comments' );` and others? Wouldn't be that the most expected behavior?

Reply:
> I was wondering why it isn't using `remove_action` against `add_action( 'init', 'register_block_core_latest_comments' );` and others? Wouldn't be that the most expected behavior?

Maybe, but there are a few complications:

- In the current stable release of WordPress, some blocks are registered immediately, not as part of an `init` hook ([example](https://github.com/WordPress/WordPress/blob/5.0.3/wp-includes/blocks/block.php#L32), addressed as part of the changes of this pull request)
- As noted in your prior review note, trying to compose these names (i.e. `register_`) dynamically is prone to being fragile
- As noted by @youknowriad 's comment, removing a filter would need to occur at an earlier priority than the default for it to work

Reply:
It's fine as is. I was mostly interested to see whether if this is technically possible without too much hassle.